app/session.server.ts

```diff
@@ -43,9 +45,47 @@ export async function getUsername(request: Request): Promise<User['username'] |
   return username;
 }
 
+export async function getEffectiveUsername(
+  request: Request
+): Promise<User['username'] | undefined> {
+  const cookie = request.headers.get('Cookie');
+  const effectiveUsername = await effectiveUsernameCookie.parse(cookie);
+  return effectiveUsername as User['username'];
+}
+
+export async function getEffectiveUser(request: Request) {
+  const username = await getUsername(request);
+  // get the effective username in the cookie at the time
+  const effectiveUsername = await getEffectiveUsername(request);
+  if (!username) {
+    return null;
+  }
+
+  // if there is no effective username let the admin be himself
+  if (!effectiveUsername || effectiveUsername === '') {
+    return await getUser(request);
```

Comment:
Here you use `getUser()` and in 69 you use `getUserByUsername()`.  Can we use the same path in both cases?  Not critical, but it feels odd that we have different code paths.